src/MongoDB/Manager.c

```diff
@@ -90,6 +90,9 @@ static bool php_phongo_manager_merge_context_options(zval* zdriverOptions TSRMLS
 #endif
 
 	php_array_unsetc(zdriverOptions, "context");
+
+	php_error_docref(NULL TSRMLS_CC, E_DEPRECATED, "The \"context\" driver option is deprecated.");
```

Comment:
I originally had this call before `zend_hash_merge` a few lines above. This resulted in an abort signal:
> [2019/10/21 09:24:53.187] 019+ php: /data/mci/8b09332019426e6646ef3090d716cfae/src/php-7.2.10/Zend/zend_hash.c:1162: zend_hash_str_del: Assertion `((ht)->gc.refcount == 1) || ((ht)->u.flags & (1<<6))' failed.
[2019/10/21 09:24:53.187] 020+
[2019/10/21 09:24:53.187] 021+ Termsig=6

I then moved it to after `zend_hash_merge` but before `php_array_unsetc`, with a similar abort being the result:
> [2019/10/21 09:24:53.187] 019+ php: /data/mci/8b09332019426e6646ef3090d716cfae/src/php-7.2.10/Zend/zend_hash.c:1162: zend_hash_str_del: Assertion `((ht)->gc.refcount == 1) || ((ht)->u.flags & (1<<6))' failed.
[2019/10/21 09:24:53.187] 020+
[2019/10/21 09:24:53.187] 021+ Termsig=6

Moving the call to its current location makes the tests pass even on evergreen. I have not been able to reproduce these failures locally.

Reply:
I assume we don't have any additional information available (e.g. full trace). I imagine the culprit here is `php_array_unsetc` rather than `zend_hash_merge`, as the unset macro calls `zend_hash_str_del` through [`zend_symtable_str_del`](https://lxr.room11.org/xref/php-src%407.3/Zend/zend_hash.h#518). Also consider that [`zend_hash_merge`](https://lxr.room11.org/xref/php-src%407.3/Zend/zend_hash.c#1998) seems to only do an add or update, but never a delete.

The `(1<<6)` flag here refers to [`HASH_FLAG_ALLOW_COW_VIOLATION`](https://lxr.room11.org/xref/php-src%407.3/Zend/zend_hash.h#41), which was introduced in https://github.com/php/php-src/commit/fd4025069d561b4196851c2e9a85b5bfdfe1465a. The actual assertion that fails is `zend_hash_str_del`'s call to [`HT_ASSERT_RC1`](https://lxr.room11.org/xref/php-src%407.2/Zend/zend_hash.c#1164), which adds a [`HASH_FLAG_ALLOW_COW_VIOLATION` check](https://lxr.room11.org/xref/php-src%407.2/Zend/zend_hash.c#27) only when PHP is compiled in debug mode. That might explain why you can't reproduce this locally.

Based on the commit that introduced this, it would seem that you could also work around this by calling `HT_ALLOW_COW_VIOLATION(Z_ARRVAL_P(zdriverOptions));`. https://github.com/php/php-src/commit/fd4025069d561b4196851c2e9a85b5bfdfe1465a does include a number of comments explaining why `HT_ALLOW_COW_VIOLATION()` is used in certain places, but I think in our case this could pertain to an additional reference to `zdriverOptions`'s hash table (remember [PHP 7's refcounts now live on the value, not the zval](https://nikic.github.io/2015/05/05/Internal-value-representation-in-PHP-7-part-1.html#zvals-in-php-7)) being created due to emitting a deprecation notice. Perhaps this is also pertinent to having set a user-defined error handler to convert this notice into an exception (as notices would not care about in-scope variables otherwise).

If you follow the rabbit hole through [`php_verror`](https://lxr.room11.org/xref/php-src%407.3/main/main.c#925), which itself is called from the `php_error_docref#` functions, you'll come to [`zend_error`](https://lxr.room11.org/xref/php-src%407.3/Zend/zend.c#1169). Therein, we see some code for the [case where a user-defined error handler has been set](https://lxr.room11.org/xref/php-src%407.3/Zend/zend.c#1286) where the current scope is captured (symbol table appears to get duplicated). If my hunch is correct, this means that we should expect `zdriverOptions`'s hash table RC to get incremented whenever a user-defined error handler is set, which is certainly the case for our tests but likewise also for most frameworks. Granted, PHP probably won't be compiled in debug mode for most users, but we shouldn't assume that.

*tl;dr*: Let's see if `HT_ALLOW_COW_VIOLATION()` fixes this and allows us to order the notice emission in the ideal location. Either way, we should add a comment explaining (a) why we call `HT_ALLOW_COW_VIOLATION` or (b) why we've deferred the call to `php_error_docref`.

----

Separate thought: let's say `$driverOptions` already had multiple references from the application (totally independent from our merging here). In that case, I expect we might also hit the RC assertion when trying to unset a field on it by calling `php_array_unsetc` from a bit of extension code. This is possibly something you can test in isolation, and combine with debugging the value of `Z_ARRVAL_P(zdriverOptions)->gc.refcount`.

Reply:
Thank you for that excellent analysis. Using `HT_ALLOW_COW_VIOLATION` fixes the error. The error does not only originate in `zend_hash_str_del`, but also in `zend_hash_merge` where it was introduced in the same commit linked above (https://github.com/php/php-src/commit/fd4025069d561b4196851c2e9a85b5bfdfe1465a#diff-fd78a0a3f78ea28c6907f907f25b908eR1858).

I wasn't able to replicate the last thought. The refcount of `driverOptions` (analysed with `xdebug_debug_zval` for convenience) returned a refcount of 4 with no error being thrown (unless adding the `php_error_docref` call). Even when explicitly duplicating the `context` element to increase its refcount to 2 I didn't get the error in question. With the information you gathered, I'd indeed assume that this is caused by a user-defined error handler leading `zend_error` copying the symbol table.

As an additional note, this entirely behaviour is related to the `$errcontext` argument in error handlers that was deprecated in PHP 7.2. It has been removed from PHP 8 already: https://github.com/php/php-src/commit/2f1f34952e9a0dfb3adcbec82ba69f4ac82b3a3d#diff-b09edfedd835ebc4491e565c147190e7L1388. I currently don't have a setup to run debug tests on PHP 8 but will try to set one up soon.